### ISSUE TITLE
How to ensure reproducibility of the results?

### ISSUE THREAD

**Issue by daidaiershidi** [yassersouri/robust_seg#1]
Thank you for proposing such an interesting work. I ran the experiments multiple times in succession and found varying results each time. Is there any way to ensure consistency in experimental outcomes? Could it be due to the 'randomness' (https://github.com/yassersouri/robust_seg/blob/5df2c629e9f77e90538896d6c601784a69626683/src/pgt.py#L382C14-L382C24) involved here?

**Comment by yassersouri** [comment]
Hi @daidaiershidi,

The inconsistency usually comes from the network training part.
Since these dataset are not that large, it is customary to train/evaluate on 4 different splits and then report the average.
This averaging does reduce the variance in the results and gives a better more reliable metric for comparison.

Actually in the `generate_pgt_hard` function that you linked to, I don't see any randomness, except when the log probabilities are random. If the log probabilities are changed, then the pseudo ground truth would change and it would then result in a change in log probs, etc.

But please keep in mind that the log probabilities inherently have randomness due to other things:
 - CUDA and GPU randomness
 - Python randomness
 - Batching randomness
 - etc.

If you haven't, I suggest that you take a look at this: https://pytorch.org/docs/stable/notes/randomness.html

But, I don't think it is a good idea (unless required for some specific reason) to remove the randomness, it is better to see it, and understand it and take it into account during evaluation.
